File: NeoVM.Interop/Types/StackItems/IntegerStackItem.cs
using NeoVM.Interop.Enums;
using NeoVM.Interop.Interfaces;
using System;
using System.Numerics;

namespace NeoVM.Interop.Types.StackItems
{
    public class IntegerStackItem : IStackItem<BigInteger>, IEquatable<IntegerStackItem>
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="data">Data</param>
        internal IntegerStackItem(ExecutionEngine engine, int data) : base(engine, new BigInteger(data), EStackItemType.Integer)
        {
            CreateNativeItem();
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="data">Data</param>
        internal IntegerStackItem(ExecutionEngine engine, long data) : base(engine, new BigInteger(data), EStackItemType.Integer)
        {
            CreateNativeItem();
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="data">Data</param>
        internal IntegerStackItem(ExecutionEngine engine, byte[] data) : base(engine, new BigInteger(data), EStackItemType.Integer)
        {
            CreateNativeItem();
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="data">Data</param>
        internal IntegerStackItem(ExecutionEngine engine, BigInteger data) : base(engine, data, EStackItemType.Integer)
        {
            CreateNativeItem();
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="handle">Handle</param>
        /// <param name="value">Raw value</param>
        internal IntegerStackItem(ExecutionEngine engine, IntPtr handle, byte[] value) :
            base(engine, new BigInteger(value), EStackItemType.Integer, handle)
        { }

        public bool Equals(IntegerStackItem other)
        {
            return other != null && other.Value.Equals(Value);
        }

        public override bool Equals(IStackItem other)
        {
            if (other is IntegerStackItem b)
                return b.Value.Equals(Value);

            return false;
        }

        protected override byte[] GetNativeByteArray()
        {
            return Value.ToByteArray();
        }
    }
}

File: src/Neo.HyperVM/IntegerStackItem.cpp
#include "IntegerStackItem.h"

bool IntegerStackItem::Equals(IStackItem* it)
{
	if (it == this) return true;

	switch (it->Type)
	{
	case EStackItemType::Integer:
	{
		auto t = (IntegerStackItem*)it;
		return this->_value.CompareTo(t->_value) == 0;
	}
	default:
	{
		int32 i1 = it->ReadByteArraySize();

		if (i1 < 0)
		{
			return false;
		}

		int32 i0 = this->ReadByteArraySize();
		byte* d0 = new byte[i0];
		byte* d1 = new byte[i1];

		i0 = this->ReadByteArray(d0, 0, i0);
		i1 = it->ReadByteArray(d1, 0, i1);

		if (i1 != i0)
		{
			delete[](d0);
			delete[](d1);
			return false;
		}

		bool ret = true;
		for (int32 x = 0; x < i0; ++x)
			if (d0[x] != d1[x])
			{
				ret = false;
				break;
			}

		delete[](d0);
		delete[](d1);
		return ret;
	}
	}
}